### PR TITLE
Fix stopping during greeting on some POPs servers

### PR DESCRIPTION
**Pull request by trueroad**
Some POPs servers (e.g. secure.plala.or.jp:995) split CAPA responses into multiple packets.
Mew stopped greeting unless the CAPA response was a single packet. This commit fixes to allow for multiple packets of CAPA responses in greetings.

### CODE REVIEW

**Comment by trueroad** [comment]
Settings for debug logging below.

```.el
(setq mew-debug t)
(setq mew-config-alist
      '((default
         (pop-server     "secure.plala.or.jp")
         (pop-ssl        native)
         (pop-user       "test@example.com")
         (pop-auth       pass))))
```

Here is a debug log before applying this PR.
The response after sending (`<=SEND=>`) `CAPA` is split into `<GREEING>` and `<CAPA>`.
The last line of the `<GREETING>` response is `UIDL`, not `.` .
The `.` is in the `<CAPA>` response but not the `<GREETING>`, so Mew stops there.

```
<TLS proto=pop, server=secure.plala.or.jp:pop3s, starttlsp=nil>
verify-level=1, network-security-level=medium, nowait=nil, tlsparams=gnutls-x509pki :priority NORMAL:%DUMBFW :hostname secure.plala.or.jp :loglevel 0 :min-prime-bits 2048 :trustfiles (/etc/pki/tls/certs/ca-bundle.crt) :crlfiles nil :keylist nil :verify-flags nil :verify-error nil :callbacks nil :priority-string NORMAL:%DUMBFW 

<=SEND=>
CAPA

<GREETING>
+OK Capability list follows
TOP
USER
RESP_CODES
PIPELINING
EXPIRE NEVER
UIDL


<=SEND=>
CAPA

<CAPA>
.


<CAPA>
+OK Capability list follows
TOP
USER
RESP_CODES
PIPELINING
EXPIRE NEVER
UIDL


<CAPA>
.
```

Here is a debug log after applying this PR.
The response after sending (`<=SEND=>`) `CAPA` is split, but both are in `<GREEING>` .
The last line of the first `<GREETING>` response is `UIDL`.
The second `<GREETING>` response has `.` .
Since the last line of the two `<GREETING>` responses is `.`, Mew proceeds to the next operation.

```
<TLS proto=pop, server=secure.plala.or.jp:pop3s, starttlsp=nil>
verify-level=1, network-security-level=medium, nowait=nil, tlsparams=gnutls-x509pki :priority NORMAL:%DUMBFW :hostname secure.plala.or.jp :loglevel 0 :min-prime-bits 2048 :trustfiles (/etc/pki/tls/certs/ca-bundle.crt) :crlfiles nil :keylist nil :verify-flags nil :verify-error nil :callbacks nil :priority-string NORMAL:%DUMBFW 

<=SEND=>
CAPA

<GREETING>
+OK Capability list follows
TOP
USER
RESP_CODES
PIPELINING
EXPIRE NEVER
UIDL


<GREETING>
.


<=SEND=>
CAPA

<CAPA>
+OK Capability list follows
TOP
USER
RESP_CODES
PIPELINING
EXPIRE NEVER
UIDL


<CAPA>
.


<=SEND=>
USER test@example.com

<USER>
+OK please send PASS command


<=SEND=>
PASS 

<PASS>
-ERR a password MUST be entered


<=SEND=>
QUIT

<QUIT>
+OK test@example.com SPOP3 server signing off.


<POP SENTINEL>
connection broken by remote peer
```

**Comment by kazu-yamamoto** [comment]
With this patch, Mew cannot negotiate my POP server correctly.
We should consider a workaround.

File: elisp/mew-pop.el [hunk 1]
```diff
@@ -39,7 +39,7 @@
 ;;;
 
 (defvar mew-pop-fsm
-  '(("greeting"      nil ("\\+OK" . "capa"))
+  '(("greeting"      t   ("\\+OK" . "capa"))
```
Comment (kazu-yamamoto) [inline elisp/mew-pop.el:42]: This changes "greeting" from a single line to multiple lines.
If the source of problem is "CAPA", this change should have no effect.

File: elisp/mew-pop.el [hunk 2]
```diff
@@ -39,7 +39,7 @@
...
     ("capa"          t   ("\\+OK" . "auth") ("-ERR" . "pswd"))
```
Comment (kazu-yamamoto) [inline elisp/mew-pop.el:43]: "capa" is defined as multiple lines.
So, I don't know why your problem happens.

**Review by kazu-yamamoto — COMMENTED** [review]
More research is necessary.

File: elisp/mew-pop.el (hunk 1 above)
Comment (trueroad) [reply elisp/mew-pop.el]: When using GnuTLS, POP greetings are not received.
So a dummy CAPA is sent and the response is used as a substitute for the greeting.
https://github.com/kazu-yamamoto/Mew/blob/d41dc93785d231f1e391ba61893aacd1331d5726/elisp/mew-pop.el#L777

File: elisp/mew-pop.el (hunk 2 above)
Comment (trueroad) [reply elisp/mew-pop.el]: My problem happens in "greeting" that is dummy CAPA instead of true CAPA.

**Comment by trueroad** [comment]
Thank you for your reviewing.

My problem seems to be only when using GnuTLS.
So, in that case, I fixed this PR to skip the dummy greeting by CAPA and begin with true CAPA immediately.

I've confirmed that all four cases below proceed to PASS.

```.el
(setq mew-debug t)
(setq mew-ssl-verify-level 0)
(setq mew-config-alist
      '((default
         (pop-server     "secure.plala.or.jp")
         (pop-port       110)
         (pop-ssl        nil)
         (pop-user       "test@example.com")
         (pop-auth       pass))
        (gnutls-ssl
         (pop-ssl        native)
         (pop-ssl-port   995))
        (stunnel-ssl
         (pop-ssl        t)
         (pop-ssl-port   995))
        (gnutls-starttls
         (pop-ssl        native)
         (pop-ssl-port   110))
        (stunnel-starttls
         (pop-ssl        t)
         (pop-ssl-port   110))))
```

**Comment by kazu-yamamoto** [comment]
FYI:
https://salt.iajapan.org/wpmu/anti_spam/admin/tech/explanation/tls-arc/

**Comment by kazu-yamamoto** [comment]
There are two types of TLS for POP:

1. STARTTLS (port 110)
2. POP over TLS/SSL (port 995)

For (2), "STARTTLS" should be listed in the response of "CAPA".

**Comment by trueroad** [comment]
いままでのパッチは破棄して佐藤さんのパッチ
https://groups.google.com/g/mew-ja/c/Z7r8K5GbKuc/m/ps_N-_gmAAAJ
を取り込みました。

佐藤さんのパッチは GnuTLS 使用時にダミーの CAPA 送信をやめて、代わりにダミーの greeting をプロセスフィルタに渡すものです。
さらに GnuTLS のリターン値から本物の greeting を得ることができるので、ダミーの greeting ではなくて本物の greeting を渡すコミットを追加してあります。

5環境

* 平文 POP （ポート 110）
* GnuTLS で POP over TLS/SSL （ポート 995）
* GnuTLS で STARTTLS （ポート 110）
* stunnel で POP over TLS/SSL （ポート 995）
* stunnel で STARTTLS （ポート 110）

いずれでも動作することを確認しております。

何か問題などありましたらご指摘いただければ幸いです。

**Comment by kazu-yamamoto** [comment]
遅くなりましたが。
かなり時間をかけて検証しました。
TLSプロトコルの設計思想に合致しており、正しい修正だと確信できたのでマージします。

**Review by kazu-yamamoto — APPROVED** [review]
LGTM

**Comment by kazu-yamamoto** [comment]
マージしました。
貢献、ありがとうございました。

**Pull request closed without merging**